src/wallet/rpcwallet.cpp

```diff
@@ -1064,6 +1065,153 @@ UniValue sendmany(const JSONRPCRequest& request)
     return wtx.GetHash().GetHex();
 }
 
+UniValue sweepprivkeys(const JSONRPCRequest& request)
+{
+    CWallet * const pwallet = GetWalletForJSONRPCRequest(request);
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp)) {
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() != 1)
+        throw std::runtime_error(
+            "sweepprivkeys {\"privkeys\": [\"bitcoinprivkey\",...], other options}\n"
+            "\nSends bitcoins controlled by private key to specified destinations.\n"
+            "\nOptions:\n"
+            "  \"privkeys\":[\"bitcoinprivkey\",...]   (array of strings, required) An array of WIF private key(s)\n"
+            "  \"label\":\"actuallabelname\"           (string, optional) Label for received bitcoins\n"
+        );
+
+    // NOTE: It isn't safe to sweep-and-send in a single action, since this would leave the send missing from the transaction history
+
+    RPCTypeCheck(request.params, {UniValue::VOBJ});
+
+    // Parse options
+    std::set<CScript> needles;
+    CCoinControl coin_control;
+    CBasicKeyStore tempKeystore;
+    CMutableTransaction tx;
+    std::string label;
+    CAmount total_in = 0;
+    for (const std::string& optname : request.params[0].getKeys()) {
+        const UniValue& optval = request.params[0][optname];
+        if (optname == "privkeys") {
+            const UniValue& privkeys_a = optval.get_array();
+            for (size_t privkey_i = 0; privkey_i < privkeys_a.size(); ++privkey_i) {
+                const UniValue& privkey_wif = privkeys_a[privkey_i];
+                std::string wif_secret = privkey_wif.get_str();
+                CKey key;
+                CPubKey pubkey;
+                ParseWIFPrivKey(wif_secret, key, &pubkey);
+
+                tempKeystore.AddKey(key);
+                CKeyID address = pubkey.GetID();
+                CScript script = GetScriptForDestination(address);
+                if (!script.empty()) {
+                    needles.insert(script);
+                }
+                script = GetScriptForRawPubKey(pubkey);
+                if (!script.empty()) {
+                    needles.insert(script);
+                }
+            }
+        } else if (optname == "label") {
+            label = AccountFromValue(optval.get_str());
+        } else {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Unrecognised option '%s'", optname));
+        }
+    }
+
+    // Ensure keypool is filled if possible
+    {
+        LOCK2(cs_main, pwallet->cs_wallet);
+
+        if (!pwallet->IsLocked()) {
+            pwallet->TopUpKeyPool();
+        }
+    }
+
+    // Reserve the key we will be using
+    CReserveKey reservekey(pwallet);
+    CPubKey pubkey;
+    if (!reservekey.GetReservedKey(pubkey)) {
+        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
```

Comment:
I don't know how, and the two tests currently "testing" it don't make logical sense.